lms/djangoapps/verify_student/utils.py

```diff
@@ -93,28 +93,29 @@ def send_verification_status_email(context):
         ))
 
 
-def most_recent_verification(photo_id_verifications, sso_id_verifications, most_recent_key):
+def most_recent_verification(photo_id_verifications, sso_id_verifications, manual_id_verifications, most_recent_key):
     """
-    Return the most recent verification given querysets for both photo and sso verifications.
+    Return the most recent verification given querysets for photo, sso and manual verifications.
 
     Arguments:
             photo_id_verifications: Queryset containing photo verifications
             sso_id_verifications: Queryset containing sso verifications
+            manual_id_verifications: Queryset containing manual verifications
             most_recent_key: Either 'updated_at' or 'created_at'
 
     Returns:
         The most recent verification.
     """
     photo_id_verification = photo_id_verifications and photo_id_verifications.first()
     sso_id_verification = sso_id_verifications and sso_id_verifications.first()
+    manual_id_verification = manual_id_verifications and manual_id_verifications.first()
 
-    if not photo_id_verification and not sso_id_verification:
-        return None
-    elif photo_id_verification and not sso_id_verification:
-        return photo_id_verification
-    elif sso_id_verification and not photo_id_verification:
-        return sso_id_verification
-    elif getattr(photo_id_verification, most_recent_key) > getattr(sso_id_verification, most_recent_key):
-        return photo_id_verification
-    else:
-        return sso_id_verification
+    verifications = [photo_id_verification, sso_id_verification, manual_id_verification]
+
+    verifications_map = {
+        verification: getattr(verification, most_recent_key)
+        for verification in verifications
+        if getattr(verification, most_recent_key, False)
+    }
+
+    return max(verifications_map, key=lambda k: verifications_map[k]) if verifications_map else None
```

Comment:
Can we use the previous code style as now it looks a bit difficult to understand the outcome.

Reply:
This refactoring was required due to the unwieldy elif statements

Reply:
The only problem I see is readability. If a piece of code is more readable if will be more maintainable.

Reply:
We would need 2^n elif statements (n is the number of verification types) ... clearly not a scalable solution

Reply:
@irfanuddinahmad I agree with @zubair-arbi about the readability of the above logic. While we will need more `elif` statements when there are additional verification types, I think the addition of the `elif` statements for `ManualVerification` is fine for now. We can consider refactoring later on if/when more verification types get added.

Reply:
@adamstankiewicz @zubair-arbi ... please review the following before I commit:

```bash
def most_recent_verification(photo_id_verifications, sso_id_verifications, manual_id_verifications, most_recent_key):
    """
    Return the most recent verification given querysets for photo, sso and manual verifications.
    Arguments:
        photo_id_verifications: Queryset containing photo verifications
        sso_id_verifications: Queryset containing sso verifications
        manual_id_verifications: Queryset containing manual verifications
        most_recent_key: Either 'updated_at' or 'created_at'
    Returns:
        The most recent verification.
    """
    photo_id_verification = photo_id_verifications and photo_id_verifications.first()
    sso_id_verification = sso_id_verifications and sso_id_verifications.first()
    manual_id_verification = manual_id_verifications and manual_id_verifications.first()

    if not photo_id_verification and not sso_id_verification and not manual_id_verification:
        return None

    elif photo_id_verification and not sso_id_verification and not manual_id_verification:
        return photo_id_verification
    elif sso_id_verification and not photo_id_verification and not manual_id_verification:
        return sso_id_verification
    elif manual_id_verification and not photo_id_verification and not sso_id_verification:
        return manual_id_verification

    elif photo_id_verification and sso_id_verification and not manual_id_verification:
        if getattr(photo_id_verification, most_recent_key) > getattr(sso_id_verification, most_recent_key):
            return photo_id_verification
        else:
            return sso_id_verification

    elif photo_id_verification and manual_id_verification and not sso_id_verification:
        if getattr(photo_id_verification, most_recent_key) > getattr(manual_id_verification, most_recent_key):
            return photo_id_verification
        else:
            return manual_id_verification

    elif sso_id_verification and manual_id_verification and not photo_id_verification:
        if getattr(sso_id_verification, most_recent_key) > getattr(manual_id_verification, most_recent_key):
            return sso_id_verification
        else:
            return manual_id_verification

    else:
        sso_key = getattr(sso_id_verification, most_recent_key)
        photo_key = getattr(photo_id_verification, most_recent_key)
        manual_key = getattr(manual_id_verification, most_recent_key)
        if sso_key > photo_key and sso_key > manual_key:
            return sso_id_verification
        elif photo_key > sso_key and photo_key > manual_key:
            return photo_id_verification
        else:
            return manual_id_verification
```